File: Java Homework1/Matrix.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <string.h>

typedef struct{
	int row;
	int col;
	double** data;
}Matrix;

Matrix NewMatrix(int row_in, int col_in){
	Matrix newM;
	int i, j;
	double randNum;
	
	newM.row = row_in;
	newM.col = col_in;
	newM.data = (double **)malloc(sizeof(double*)*row_in);
	for(i = 0; i < row_in; i++){
		newM.data[i] = (double *)malloc(sizeof(double)*col_in);
	}
	for(i = 0; i < row_in; i++){
		for(j = 0; j < col_in; j++){
			srand(time(NULL));
			randNum = (rand()/((double)RAND_MAX+1))*RAND_MAX + RAND_MAX;
			newM.data[i][j] = randNum;
		}
	}
	return newM;	
}

int main(){
	int sttRow, endRow, sttCol, endCol;
	int i, j;
	int p, q, k;
	int size;
	Matrix m, n, r;
	FILE *pFile;
	struct timeval tvStart,tvEnd;
	long elapsedTime;
	
	printf(">>>>please input row-length range and columb-length range\n");
	printf(">>>>all values must be larger than zero\n");
	printf(">>>>ie: row start point: 2; row end point: 10\n");
	printf(">>>>    then the row-length range is from 2 to 10\n");
	printf(">>>>row start point: ");
	scanf("%d", &sttRow);
	printf(">>>>row end point: ");
	scanf("%d", &endRow);
	printf(">>>>col start point: ");
	scanf("%d", &sttCol);
	printf(">>>>col end point: ");
	scanf("%d", &endCol);
	
	pFile = fopen("ResultsByC.txt", "w+");
	fprintf(pFile, "Size\tElapsedTime(in Msec)\n");
	for(i = sttRow; i <= endRow; i++){
		for(j = sttCol; j <=endCol; j++){
			m = NewMatrix(i, j);
			n = NewMatrix(j, i);
			r = NewMatrix(i, i);
			gettimeofday(&tvStart, NULL);
			for(p = 0;p < i;p++){
				for(q = 0;q < i; q++){
					for(k = 0; k < j; k++){
						r.data[p][q] += (m.data[p][k]*n.data[k][q]);
					}
				}
			}
			gettimeofday(&tvEnd, NULL);
			elapsedTime = (long)(tvEnd.tv_sec - tvStart.tv_sec)*1000 + (long)(tvEnd.tv_usec - tvStart.tv_usec)/1000;
			
			size = i*j;
			fprintf(pFile, "%d\t%ld\n", size, elapsedTime);
			free(m.data);
			free(n.data);
			free(r.data);
		}
	}
	fclose(pFile);
}

File: Java Homework1/Matrix.java
import java.util.*;
import java.io.*;

class Matrix{
	private int row, col;
	private double[][] data;
	
	public Matrix(int row_in, int col_in){
		row = row_in;
		col = col_in;
		data = new double[row][col];
	}
	
	public static void main(String[] args){
		System.out.print(">>>>please input row-length range and column-length range\n");
		System.out.print(">>>>ie: 1-10 2-20\n");
		System.out.print(">>>>all values must be larger than zero\n");
		System.out.print(">>>>");
		String input = ParserUtils.getKeyInput();
		String[] inputToken = ParserUtils.getTokens(input);
		if(inputToken.length != 2){
			ErrorMsg("- error: too many/few input values");
		}
		for(int i = 0; i < 2; i++){
			if(!inputToken[i].matches("\\d+-\\d+")){
				ErrorMsg("- error: invalid inputs");
			}
		}
		String[] stringData;
		stringData = inputToken[0].split("-");
		int rowSttPoint = Integer.parseInt(stringData[0]);
		int rowEndPoint = Integer.parseInt(stringData[1]);
		stringData = inputToken[1].split("-");
		int colSttPoint = Integer.parseInt(stringData[0]);
		int colEndPoint = Integer.parseInt(stringData[1]);
		try{
			BufferedWriter outFile = new BufferedWriter(new FileWriter("ResultsByJava.txt"));
			outFile.write("Size\tElapsedTime(in Msec)\n");
			for(int i = rowSttPoint; i <= rowEndPoint; i++){
				for(int j = colSttPoint; j <= colEndPoint; j++){
					Matrix m = NewMatrix(i, j);
					Matrix n = NewMatrix(j, i);
					Matrix result = new Matrix(i, i);
					long start = System.currentTimeMillis();
					for(int p = 0; p < i; p++){
						for(int q = 0; q < i; q++){
							for(int r = 0; r < j; r++){
								result.data[p][q] += (m.data[p][r]*n.data[r][q]);
							}
						}
					}
					long end = System.currentTimeMillis();
					long elapsedTime = end - start;
					int size = i*j;
					outFile.write(Integer.toString(size));
					outFile.write("\t");
					outFile.write(Long.toString(elapsedTime));
					outFile.write("\n");
				}
			}
			outFile.close();
		}catch(IOException e){System.out.println("- error while writing file: "+e);}
	}
	
	private static void ErrorMsg(String msg){
		System.out.println(msg);
		System.exit(0);
	}
	
	public static Matrix NewMatrix(int row, int col){
		Matrix newM = new Matrix(row, col);
		for(int i = 0; i < row; i++)
			for(int j = 0; j < col; j++){
				Random randomGenerator = new Random();
				double ranData = randomGenerator.nextDouble()*100;
				newM.data[i][j] = ranData;
			}
		return newM;
	}
}